### PR TITLE
Add the auto_find_batch_size capability from Accelerate into Trainer

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR introduces the `find_executable_batch_size` decorator into `Trainer`, so the training loop is repeated if a CUDA OOM is reached, lowering the batch size. 

The API looks as so:

```python
trainer = Trainer()
trainer.train(auto_find_batch_size=True)
```
By default it is False, and requires `Accelerate` be installed to use. 

Fixes # (issue)

Partially solves https://github.com/huggingface/transformers/issues/16987

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sgugger

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -96,6 +96,7 @@
 # 2. once modified, run: `make deps_table_update` to update src/transformers/dependency_versions_table.py
 _deps = [
     "Pillow",
+    "accelerate",
```
Comment (sgugger) [inline setup.py:99]:
```suggestion
    "accelerate>=0.7.1",
```
We need the latest release for this feature.

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -1024,6 +1024,24 @@ def test_resume_training_with_randomness(self):
...
+
+        config = RegressionModelConfig(a=0, b=2)
+        model = RegressionRandomPreTrainedModel(config)
+
+        tmp_dir = self.get_auto_remove_tmp_dir()
+        args = RegressionTrainingArguments(tmp_dir, save_steps=5, learning_rate=0.1, per_device_train_batch_size=1024)
+        trainer = Trainer(model, args, train_dataset=train_dataset, eval_dataset=eval_dataset)
+        with self.assertRaises(RuntimeError):
+            trainer.train(auto_find_batch_size=False)
+        trainer.train(auto_find_batch_size=True)
```
Comment (sgugger) [inline tests/trainer/test_trainer.py:1043]: The model has two parameters here, so even with a batch size of 1024, I don't think we will get OOM ;-)

File: tests/trainer/test_trainer_utils.py [hunk 3]
```diff
@@ -420,3 +421,37 @@ def test_shard_sampler(self):
...
+        def mock_training_loop_function(batch_size):
+            nonlocal batch_sizes
+            batch_sizes.append(batch_size)
+            if batch_size > 16:
+                raise RuntimeError("CUDA out of memory.")
+
+        mock_training_loop_function()
+        self.assertEqual(batch_sizes, [64, 32, 16])
+
+    def test_executable_batch_size_no_search(self):
```
Comment (sgugger) [inline tests/trainer/test_trainer_utils.py:439]: This needs the `require_accelerate` decorator too.

File: tests/trainer/test_trainer_utils.py [hunk 4]
```diff
@@ -420,3 +421,37 @@ def test_shard_sampler(self):
...
+
+        @find_executable_batch_size(starting_batch_size=64, auto_find_batch_size=False)
+        def mock_training_loop_function(batch_size):
+            nonlocal batch_sizes
+            batch_sizes.append(batch_size)
+
+        mock_training_loop_function()
+        self.assertEqual(batch_sizes, [64])
+
+    def test_executable_batch_size_with_error(self):
```
Comment (sgugger) [inline tests/trainer/test_trainer_utils.py:450]: Here too.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this! Left a few comments to make the PR a bit better :-)

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -1024,6 +1019,40 @@ def test_resume_training_with_randomness(self):
         self.assertAlmostEqual(a, a1, delta=1e-8)
         self.assertAlmostEqual(b, b1, delta=1e-8)
 
+    @require_torch_non_multi_gpu
```
Comment (sgugger) [inline tests/trainer/test_trainer.py:1022]: Should be a slow test.

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1201,6 +1205,19 @@ def train(
...
             self.model_wrapped = self.model
 
+        return find_executable_batch_size(
+            self._inner_training_loop, self._train_batch_size, args.auto_find_batch_size
+        )(
+            args=args,
+            resume_from_checkpoint=resume_from_checkpoint,
+            trial=trial,
+            ignore_keys_for_eval=ignore_keys_for_eval,
+        )
```
Comment (sgugger) [inline src/transformers/trainer.py:1215]: This is a bit hard to parse with formatting. How about:
``
```suggestion
        inner_training_loop = find_executable_batch_size(
            self._inner_training_loop, self._train_batch_size, args.auto_find_batch_size
        )
        return inner_training_loop(
            args=args,
            resume_from_checkpoint=resume_from_checkpoint,
            trial=trial,
            ignore_keys_for_eval=ignore_keys_for_eval,
        )
```

**Review by sgugger — APPROVED** [review]
Thanks for all the work on this. Pinging @LysandreJik to have a second set of eye :-)

**Comment by muellerzr** [comment]
@stas00 I'm getting a test failure on the metrics:

```python
 tests/trainer/test_trainer.py:1426: in check_mem_metrics
     metrics = trainer.train().metrics
 src/transformers/trainer.py:1215: in train
     ignore_keys_for_eval=ignore_keys_for_eval,
 src/transformers/trainer.py:1571: in _inner_training_loop
     self._memory_tracker.stop_and_update_metrics(metrics)
 src/transformers/trainer_utils.py:536: in stop_and_update_metrics
     stage = self.derive_stage()
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 
 self = <transformers.trainer_utils.TrainerMemoryTracker object at 0x7f929f787e10>
 
     def derive_stage(self):
         """derives the stage/caller name automatically"""
         caller = inspect.currentframe().f_back.f_back.f_code.co_name
         if caller in self.stages:
             return self.stages[caller]
         else:
             raise ValueError(
 >               f"was called from {caller}, but only expect to be called from one of {self.stages.keys()}"
             )
 E           ValueError: was called from _inner_training_loop, but only expect to be called from one of dict_keys(['__init__', 'train', 'evaluate', 'predict'])
```

Any advice on how to approach a solution?

**Comment by stas00** [comment]
This will overcome the problem:

```
diff --git a/src/transformers/trainer_utils.py b/src/transformers/trainer_utils.py
index 22b44a2f0..d4c523249 100644
--- a/src/transformers/trainer_utils.py
+++ b/src/transformers/trainer_utils.py
@@ -356,6 +356,7 @@ class TrainerMemoryTracker:
     stages = {
         "__init__": "init",
         "train": "train",
+        "_inner_training_loop": "train",
         "evaluate": "eval",
         "predict": "test",
     }
```

**Review by LysandreJik — APPROVED** [review]
Great, looks good to me!

**Comment by LysandreJik** [comment]
Please make sure all tests pass after resolving conflicts and before merging!

**Comment by JohnGiorgi** [comment]
Any chance similar functionality could be supported for inference? 🙏
